IoT is internet of thing

With the advent of IoT, the attack surfaces on any given target have increased exponentially.


In this project we will build a machine learning model to classify netwrok IoT devices by type.

A dataset has 2 types which are iot_train.csv and iot_test.csv

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
%cd /content/drive/My Drive/Colab Notebooks

/content/drive/My Drive/Colab Notebooks


In [4]:
# import pandas and os and read in the training and testing data
import pandas as pd
import os
training_data = pd.read_csv("iot_devices_train.csv")
testing_data = pd.read_csv("iot_devices_test.csv")

In [5]:
print(training_data)

     ack  ack_A  ack_B  ...  suffix_is_org  suffix_is_unresolved  device_category
0      9      5      5  ...              0                     0  security_camera
1      9      5      5  ...              0                     0  security_camera
2      9      5      5  ...              0                     0  security_camera
3      9      5      5  ...              0                     0  security_camera
4      9      5      5  ...              0                     0  security_camera
..   ...    ...    ...  ...            ...                   ...              ...
995    0      0      0  ...              0                     0           socket
996    0      0      0  ...              0                     0           socket
997    0      0      0  ...              0                     0           socket
998    0      0      0  ...              0                     0           socket
999    0      0      0  ...              0                     0           socket

[1000 rows x 29

In [6]:
print(testing_data)

     ack  ack_A  ack_B  ...  suffix_is_org  suffix_is_unresolved  device_category
0     38     20     18  ...              0                     0     baby_monitor
1     38     20     18  ...              0                     0     baby_monitor
2     38     20     18  ...              0                     0     baby_monitor
3     38     20     18  ...              0                     0     baby_monitor
4     41     20     21  ...              0                     0     baby_monitor
..   ...    ...    ...  ...            ...                   ...              ...
895  264    116    148  ...              0                     0            watch
896   24     13     11  ...              0                     0            watch
897   20      9     11  ...              0                     0            watch
898   30     14     16  ...              0                     0            watch
899  294    147    147  ...              0                     0            watch

[900 rows x 298

That means the data contains 298 features

In [7]:
# We will create a training and testing dataset where the target is the device category:
x_train,y_train = (
    training_data.loc[:, training_data.columns != "device_category"].values,
    training_data["device_category"],
)

x_test,y_test = (
    testing_data.loc[:, testing_data.columns != "device_category"].values,
    testing_data["device_category"],
)

The device categories are security camera, TV, smoke detector, thermostat, water sensor, watch, baby monitor, motion sensor, lights and socket

In [8]:
# We need to encode the class categories into numerical form:
from sklearn import preprocessing
le= preprocessing.LabelEncoder()
le.fit(training_data["device_category"].unique())
y_train_encoded = le.transform(y_train)
y_test_encoded = le.transform(y_test)

In [9]:
# We will instantiate XGBoost classifier
from xgboost import XGBClassifier
model = XGBClassifier()

In [10]:
# We need to train and test the xgboost classifier
model.fit(x_train, y_train_encoded)
model.score(x_test, y_test_encoded)

0.6622222222222223

An important motivation for this project is that we cant reply on the IP address as an identifier of the device, since this value can be spoofed. Consequently we would like to analyze the traffic's high-level data, that is the metadata and traffic statistics, rather than content to determine wheterh the device belongs to the network. 

1.   Step 1:We begin by reading in the training and testing dataset on to featurize these and perform a quick data exploration step by observing the classification labels
2.   Step 2: We feed these into our classifier.

1.   Step 3: we convert these categorical labels into numerical ones to be used to train our machine learning classifier.

1.   Step 4: We having featurized the data

1.   Step 5: We instantiate train and test an xgboost classifier, obtaining a score of 0.66 on the testing set.


The baseline of randomly guessing between 10 would yeild an accuracy of 0.1. The XGBoost classifier trained here attains an accuracy of 0.66 suggesting that it is indeed a promising approach to classify IoT devices successfully based on high-level traffic data.



